crates/ruff_server/src/edit/range.rs

```diff
@@ -83,9 +91,50 @@ impl RangeExt for lsp_types::Range {
 impl ToRangeExt for TextRange {
     fn to_range(&self, text: &str, index: &LineIndex, encoding: PositionEncoding) -> types::Range {
         types::Range {
-            start: offset_to_position(self.start(), text, index, encoding),
-            end: offset_to_position(self.end(), text, index, encoding),
+            start: source_location_to_position(&offset_to_source_location(
+                self.start(),
+                text,
+                index,
+                encoding,
+            )),
+            end: source_location_to_position(&offset_to_source_location(
+                self.end(),
+                text,
+                index,
+                encoding,
+            )),
+        }
+    }
+
+    fn to_notebook_range(
+        &self,
+        text: &str,
+        source_index: &LineIndex,
+        notebook_index: &NotebookIndex,
+        encoding: PositionEncoding,
+    ) -> (usize, types::Range) {
+        let start = offset_to_source_location(self.start(), text, source_index, encoding);
+        let cell = notebook_index.cell(start.row);
+        let mut end = offset_to_source_location(self.end(), text, source_index, encoding);
+
+        // weird edge case here - if the end of the range is where the newline after the cell got added (making it 'out of bounds')
+        // we need to move it one character back (which should place it at the end of the last line).
+        if notebook_index.cell(end.row) != cell {
+            end = offset_to_source_location(
+                self.end().checked_sub(1.into()).unwrap_or_default(),
+                text,
+                source_index,
+                encoding,
+            );
         }
```

Comment:
Yeah, I remember this. It's because the cell content are concatenated by a newline but the newline doesn't actually exists in the original source code. So, when an edit tries to remove the entire line, the range includes the newline. This means the edit range goes out of the cell. I thought we fixed this. Let me check the code.

Reply:
Here is where I fix it:

https://github.com/astral-sh/ruff/blob/75f302298c2b6d481a466c463389ce770aa38f6b/crates/ruff_linter/src/message/json.rs#L110-L152

Related PR for additional context.

Reply:
Thanks! So it looks like this is just something we'll have to deal with.